File: app/src/main/java/net/mbonnin/arcanetracker/ui/overlay/Overlay.java
package net.mbonnin.arcanetracker.ui.overlay;

import android.content.Context;
import android.graphics.Color;
import android.graphics.Typeface;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.CheckBox;
import android.widget.LinearLayout;
import android.widget.TextView;

import net.mbonnin.arcanetracker.BuildConfig;
import net.mbonnin.arcanetracker.HDTApplication;
import net.mbonnin.arcanetracker.R;
import net.mbonnin.arcanetracker.Settings;
import net.mbonnin.arcanetracker.ViewManager;
import net.mbonnin.arcanetracker.ui.overlay.view.MainViewCompanion;

public class Overlay {
    private static Overlay sOverlay;

    public static Overlay get() {
        if (sOverlay == null) {
            sOverlay = new Overlay();
        }

        return sOverlay;
    }

    public void show() {
        MainViewCompanion.Companion.get().setState(MainViewCompanion.Companion.getSTATE_PLAYER(), false);
        MainViewCompanion.Companion.get().show(true);

        Context context = HDTApplication.Companion.getContext();

        int previousVersion = Settings.INSTANCE.get(Settings.INSTANCE.getVERSION(), 0);
        Settings.INSTANCE.set(Settings.INSTANCE.getVERSION(), BuildConfig.VERSION_CODE);

        if (Settings.INSTANCE.get(Settings.INSTANCE.getSHOW_CHANGELOG(), true)
                && previousVersion > 0
                && previousVersion < BuildConfig.VERSION_CODE) {
            View view = LayoutInflater.from(context).inflate(R.layout.whats_new, null);
            ViewManager.Params params = new ViewManager.Params();
            params.setX(ViewManager.Companion.get().getWidth() / 4);
            params.setY(ViewManager.Companion.get().getHeight() / 16);
            params.setW(ViewManager.Companion.get().getWidth() / 2);
            params.setH(7 * ViewManager.Companion.get().getHeight() / 8);

            CheckBox checkBox = view.findViewById(R.id.checkbox);
            checkBox.setChecked(true);

            int v = BuildConfig.VERSION_CODE;

            LinearLayout linearLayout = view.findViewById(R.id.changelog);
            int foundChangelog = 0;
            while (v > previousVersion) {
                int id = context.getResources().getIdentifier("changelog_" + v, "string", context.getPackageName());
                if (id > 0) {
                    String c = context.getString(id);
                    TextView textView = new TextView(context);
                    textView.setTextSize(24);
                    textView.setTextColor(Color.WHITE);
                    textView.setText(context.getString(R.string.cVersion, Integer.toString(v)));
                    textView.setTypeface(Typeface.DEFAULT_BOLD);

                    linearLayout.addView(textView);
                    foundChangelog++;

                    String items[] = c.split("\n");
                    for (String item: items) {
                        String s[] = item.split("\\{-\\}");
                        textView = new TextView(context);
                        textView.setTextSize(16);
                        textView.setTextColor(Color.WHITE);
                        textView.setText("• " + s[0]);
                        linearLayout.addView(textView);

                        if (s.length > 1) {
                            textView = new TextView(context);
                            textView.setTextSize(12);
                            textView.setTextColor(Color.WHITE);
                            textView.setText(s[1]);
                            linearLayout.addView(textView);
                        }
                    }
                }

                v--;
            }
            if (foundChangelog > 0) {
                view.findViewById(R.id.ok).setOnClickListener(unused -> {
                    ViewManager.Companion.get().removeView(view);
                    Settings.INSTANCE.set(Settings.INSTANCE.getSHOW_CHANGELOG(), checkBox.isChecked());
                });
                ViewManager.Companion.get().addModalView(view, params);
            }
        }

        Onboarding.INSTANCE.start();
    }

    public void hide() {
        ViewManager.Companion.get().removeAllViews();
    }
}

File: app/src/main/kotlin/net/mbonnin/arcanetracker/ui/overlay/Overlay.kt
package net.mbonnin.arcanetracker.ui.overlay

import android.view.LayoutInflater
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import net.hearthsim.hslog.PossibleSecret
import net.hearthsim.hsmodel.enum.CardId
import net.mbonnin.arcanetracker.*
import net.mbonnin.arcanetracker.ui.overlay.view.MainViewCompanion
import net.mbonnin.arcanetracker.ui.overlay.view.TopDrawerCompanion
import net.mbonnin.arcanetracker.ui.overlay.view.WhatsNewCompanion

object Overlay {
    fun show() {
        MainViewCompanion.get().setState(MainViewCompanion.STATE_PLAYER, false)
        MainViewCompanion.get().show(true)

        val context = ArcaneTrackerApplication.context

        val previousVersion = Settings[Settings.VERSION, 0]
        Settings[Settings.VERSION] = BuildConfig.VERSION_CODE

        if (Settings[Settings.SHOW_CHANGELOG, true]
                && previousVersion > 0
                && previousVersion < BuildConfig.VERSION_CODE) {
            val view = LayoutInflater.from(context).inflate(R.layout.whats_new, null)
            WhatsNewCompanion(view, previousVersion)
        }

        Onboarding.start()

        setAlphaProgress(getAlphaProgress())
        setButtonWidth(getButtonWidth())

        /*val possibleSecrets = listOf(
                PossibleSecret(CardId.COUNTERSPELL, 1),
                PossibleSecret(CardId.SPELLBENDER, 1),
                PossibleSecret(CardId.ICE_BARRIER, 0),
                PossibleSecret(CardId.DUPLICATE, 1),
                PossibleSecret(CardId.MIRROR_ENTITY, 1),
                PossibleSecret(CardId.EFFIGY, 0)
        )

        GlobalScope.launch(Dispatchers.Main) {
            delay(4000)
            var i = 3
            while (true) {
                if (i % 5 == 0) {
                    MainViewCompanion.get().onSecrets(
                            emptyList()
                    )

                } else {
                    MainViewCompanion.get().onSecrets(
                            possibleSecrets.shuffled()
                    )
                }
                delay(2000)
                i++
            }
        }*/
    }

    fun setAlphaProgress(progress: Int) {
        Settings.set(Settings.ALPHA, progress)
        MainViewCompanion.get().setAlpha(progress)
    }

    fun getAlphaProgress(): Int {
        return Settings.get(Settings.ALPHA, 100)
    }

    fun setButtonWidth(buttonWidth: Int) {
        Settings.set(Settings.BUTTON_WIDTH, buttonWidth - Utils.dpToPx(8))
        MainViewCompanion.get().setButtonWidth(buttonWidth)
    }

    fun getButtonWidth(): Int {
        var w = Settings.get(Settings.BUTTON_WIDTH, 0) + Utils.dpToPx(8) // when adding the tutorial, I made the button slightly smaller than what they used to be
        if (w < minButtonWidth || w >= maxButtonWidth) {
            val dp = if (Utils.is7InchesOrHigher) 50 else 38
            w = Utils.dpToPx(dp)
        }

        return w
    }

    val maxButtonWidth = Utils.dpToPx(75)

    val minButtonWidth = Utils.dpToPx(20)

    fun hide() {
        ViewManager.get().removeAllViews()
    }
}
